### ISSUE TITLE
a way to run this model on DGX Spark GB10 since its basicly arm64

### ISSUE THREAD

**Issue by saikanov** [fishaudio/fish-speech#1262]
### Self Checks

- [x] I have thoroughly reviewed the project documentation (installation, training, inference) but couldn't find any relevant information that meets my needs. [English](https://speech.fish.audio/) [中文](https://speech.fish.audio/zh/) [日本語](https://speech.fish.audio/ja/) [Portuguese (Brazil)](https://speech.fish.audio/pt/)
- [x] I have searched for existing issues [search for existing issues]([https://github.com/langgenius/dify/issues](https://github.com/fishaudio/fish-speech/issues)), including closed ones.
- [x] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/fishaudio/fish-speech/issues/515)).
- [x] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [x] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell us your story.

yes, since my device is not standar amd64 i got problem when using arm64 device like DGX Spark GB10

### 2. What is your suggested solution?

i tried blindly using this docker command

docker build     --platform linux/arm64     -f docker/Dockerfile     --build-arg BACKEND=cuda     --build-arg CUDA_VER=12.9.0     --build-arg UV_EXTRA=cu129     --target webui     -t fish-speech-webui:cuda .

and got this error

 => [app-base  5/10] WORKDIR /app                                                                                                                                              0.0s
 => [app-base  6/10] COPY --chown=1000:1000 pyproject.toml uv.lock README.md ./                                                                                                0.1s
 => ERROR [app-base  7/10] RUN --mount=type=cache,target=/tmp/uv-cache,uid=1000,gid=1000     uv python pin 3.12     && uv sync --extra cu129 --frozen --no-install-project     0.3s
------
 > [app-base  7/10] RUN --mount=type=cache,target=/tmp/uv-cache,uid=1000,gid=1000     uv python pin 3.12     && uv sync --extra cu129 --frozen --no-install-project:
0.244 Pinned `.python-version` to `3.12`
0.251 Using CPython 3.12.3 interpreter at: /usr/bin/python3.12
0.251 Creating virtual environment at: .venv
0.281 error: Distribution `triton==3.4.0 @ registry+https://pypi.org/simple` can't be installed because it doesn't have a source distribution or wheel for the current platform
0.281
0.281 hint: You're on Linux (`manylinux_2_39_aarch64`), but `triton` (v3.4.0) only has wheels for the following platforms: `manylinux_2_27_x86_64`, `manylinux_2_28_x86_64`; consider adding your platform to `tool.uv.required-environments` to ensure uv resolves to a version with compatible wheels
------
Dockerfile:211
--------------------
 210 |     # Use a generic cache path that works regardless of username
 211 | >>> RUN --mount=type=cache,target=/tmp/uv-cache,uid=${USER_UID},gid=${USER_GID} \
 212 | >>>     uv python pin ${PY_VER} \
 213 | >>>     && uv sync --extra ${UV_EXTRA} --frozen --no-install-project
 214 |
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c uv python pin ${PY_VER}     && uv sync --extra ${UV_EXTRA} --frozen --no-install-project" did not complete successfully: exit code: 2

before i spend time, can i know that this model could run on my device?

### 4. Can you help us with this feature?

- [x] I am interested in contributing to this feature.

**Comment by nic-nvidia** [comment]
Any luck?

**Comment by Dinsmoor** [comment]
**Yes... got it working on the spark but performance is bad**

Hey dude could you format your bug reports more cleanly next time, I can't even read it, use code blocks mr ai enthusiast or your own local ai to help you reformat things.

wanted this to run (https://speech.fish.audio/install/#use-docker-compose):
`tyler@spark:~/fish-speech$ COMPILE=1 docker compose --profile server up`

would get this error:
```
tyler@spark:~/fish-speech$ COMPILE=1 docker compose --profile server up
[+] Building 257.3s (17/21)                                                                                                                                 

~~~~ build stuff ~~~~~

 => ERROR [app-base  7/10] RUN --mount=type=cache,target=/tmp/uv-cache,uid=1000,gid=1000     uv python pin 3.12     && uv sync --extra cu129 --frozen  0.3s 
------                                                                                                                                                      
 > [app-base  7/10] RUN --mount=type=cache,target=/tmp/uv-cache,uid=1000,gid=1000     uv python pin 3.12     && uv sync --extra cu129 --frozen --no-install-project:
0.222 Pinned `.python-version` to `3.12`
0.230 Using CPython 3.12.3 interpreter at: /usr/bin/python3.12
0.230 Creating virtual environment at: .venv
0.261 error: Distribution `triton==3.4.0 @ registry+https://pypi.org/simple` can't be installed because it doesn't have a source distribution or wheel for the current platform
0.261 
0.261 hint: You're on Linux (`manylinux_2_39_aarch64`), but `triton` (v3.4.0) only has wheels for the following platforms: `manylinux_2_27_x86_64`, `manylinux_2_28_x86_64`; consider adding your platform to `tool.uv.required-environments` to ensure uv resolves to a version with compatible wheels
------
[+] up 0/1
 ⠙ Image fish-speech-server Building                                                                                                                  257.3s
Dockerfile:211

--------------------

 210 |     # Use a generic cache path that works regardless of username

 211 | >>> RUN --mount=type=cache,target=/tmp/uv-cache,uid=${USER_UID},gid=${USER_GID} \

 212 | >>>     uv python pin ${PY_VER} \

 213 | >>>     && uv sync --extra ${UV_EXTRA} --frozen --no-install-project

 214 |     

--------------------

failed to solve: process "/bin/sh -c uv python pin ${PY_VER}     && uv sync --extra ${UV_EXTRA} --frozen --no-install-project" did not complete successfully: exit code: 2
```

And tried all sorts of stuff with the dockerfile to conditionally get the right wheels and it was just a mess, just like all python dependancy management system under the sun since forever.

Ended up adding the wheels directly from pytorch for aarch64 to the uv.lock file (on repo commit 3dd1f85):
```
      5416 +    { url = "https://download.pytorch.org/whl/triton-3.4.0-cp310-cp310-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl", hash = "sha256
           +:86eb0cb7634328cd577fa51b59338f1af4199cf9ab62385ac42a82783d9116a4", size = 145056932 },                                                  
      5417 +    { url = "https://download.pytorch.org/whl/triton-3.4.0-cp311-cp311-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl", hash = "sha256
           +:dd8d32d337f58124fe254043d282a4bc133c2aca7989bc6358b2c546d2121513", size = 145152665 },                                                  
      5418 +    { url = "https://download.pytorch.org/whl/triton-3.4.0-cp312-cp312-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl", hash = "sha256
           +:00ea4a1c15f079bbf49ff02c312210fdb8ff05b1503ea30812610e2a3b171f3b", size = 145173818 },     
```

Which allowed it to build, because the lock file was generated on x86_64 and only contained x86_64 PyPI wheels for triton. PyPI has no aarch64 triton wheels, and the pytorch wheel index (download.pytorch.org/whl/cu129) does have them but doesn't list them in its simple index. A recurring theme on the spark.

Idk if there's a more elegant way but that's how I eventually forced it to compose.

Later though, both the references and checkpoints folder end up owned by root for some reason and I had to change them back to my user with chown.

` tyler@spark:~/fish-speech$ sudo chown tyler:tyler checkpoints/ references/`

And then, we are missing the actual checkpoint since fish-speech doesn't autodownload anything, so we have to use hugginface-cli to download:

`curl -LsSf https://hf.co/cli/install.sh | bash` and it should be added to your user's .local/bin and be generally avaliable in your terminal but if not add it to your path like `PATH=$PATH:/home/tyler/.local/bin` in your .bashrc

Then from your working dir, (set the HF_TOKEN envvar from huggingface for faster download)

`hf download fishaudio/s2-pro --local-dir checkpoints/s2-pro` 

Then re-running...

```
tyler@spark:~/fish-speech$ COMPILE=1 docker compose --profile server up
~~~ we get ~~~
server-1  |     Found GPU0 NVIDIA GB10 which is of cuda capability 12.1.
server-1  |     Minimum and Maximum cuda capability supported by this version of PyTorch is
server-1  |     (8.0) - (12.0)
```

**DANG!!!!**

However, running it, omitting the `COMPILE=1` env var, it does actually start and we get:

```
server-1  | 2026-05-12 02:46:57.900 | INFO     | tools.api_server:initialize_app:99 - Startup done, listening server at http://0.0.0.0:8080

```

Yay! But sucks about the, what, quoted 10x performance loss?

So, to test it out...
```
curl -s -X POST http://spark.local:8080/v1/tts \
    -H "Content-Type: application/json" \
    -d '{"text": "she got me thinkin about those beans."}' \
    -o beans.wav
```
^this works, it's just very very slow because of the lack of the compile flag and takes a whole .... **38 SECONDS!!!** AHHHHHHHHHHH!!!!

Would LOVE some repo owner help with making the aarch64/blackwell GB10  support with cm_121a or better for these torch and triton versions. Bump things you know?

Hope this helps you and others, and hey @leng-yue can someone from your team take a look at this and bump up the versions? Thanks.

Beans:

[beans.wav](https://github.com/user-attachments/files/27619011/beans.wav)

**Comment by Dinsmoor** [comment]
^ bumping the version got these results from
non-compile on torch 2.8.0:
```
tyler@dinsmoor:~$ time curl -s -X POST http://spark.local:8080/v1/tts     -H "Content-Type: application/json"     -d '{"text": "[angry] i hate pissing! Every time I drink water, I do it for what? Just to piss?"}'     -o pissing.wav
real    4m45.103s
```

to with 2.10.0 with COMPILE=1 :
```
tyler@dinsmoor:~$ time curl -s -X POST http://spark.local:8080/v1/tts     -H "Content-Type: application/json"     -d '{"text": "[angry] i hate pissing! Every time I drink water, I do it for what? Just to piss?"}'     -o pissing.wav

real    0m13.128s
```

check out https://github.com/fishaudio/fish-speech/pull/1285 or my little fork at https://github.com/Dinsmoor/fish-speech/ for your dgx spark

So a pretty great improvement and no need to manually edit the lock file anymore and get way better performance.

Again, works on my machine, would be great for you to test as well.

**Comment by nic-nvidia** [comment]
Tested this on a GB10 and saw ~6.5-7x speedup. Thanks!

Tested with NVIDIA’s PyTorch `2.10.0a0+b4e4ee8` with and without `--compile`.

| commit | container base | mode | run 1 | run 2 |
|---|---|---:|---:|---:|
| `a391467946cc9c396c025fc69eba10ef0b38332e` | `nvcr.io/nvidia/pytorch:25.12-py3` | no compile | 226.1s | 237.7s |
| `f3e2c6a83772cd4987c82914b949953553ee1464` | `nvcr.io/nvidia/pytorch:25.12-py3` | `--compile` | 34.7s | 34.5s |

`nvidia-smi` shows the GPU is used but suspiciously low power draw of 38w, i suspect there's still room for improvement.

**Comment by Dinsmoor** [comment]
@nic-nvidia feel free to submit your own improvements for lack of gpu utilization if you make any to my fork (can always be pushed upstream), but idk where the owner(s) of this repo are and if they are accepting PRs. There's 12 outstanding PRs in their repo. I know it's only been a couple months since one of their members/contributors commented something somewhere but this is a fast moving field 🤡 

Maybe that makes ME the goober! I am a big fan of writing code to where it only needs to change if a new platform comes up. I hate this ecosystem of lack of backwards compatibility, it's like the entire field is someone's personal dev environment where they spaghetti everywhere.

Anyway, the latest torch version is... uh 2.11 so another minor from what I had done, but I don't know if that would improve things further, and I didn't check if all the dependencies would resolve to that version. You're welcome to try it out and would be pretty easy! I won't look at it again until this becomes a bottleneck, I just wanted to get this piece together for a larger project of mine.

**Comment by Dinsmoor** [comment]
Then ping the project maintainers then the ball is in their court
